Add Blazor support to an ASP.NET Core app (#30791)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/integration.md

* For interactive components that are directly routable from user requests, see the [Use routable components](#use-routable-components) section. Follow this guidance when visitors should be able to make an HTTP request in their browser for a component with an [`@page`](xref:mvc/views/razor#page) directive.

## Add Blazor support to an ASP.NET Core app

This section covers adding Blazor support to an ASP.NET Core app:

* [Add static server Razor component rendering](#add-static-server-razor-component-rendering)
* [Enable interactive server rendering](#enable-interactive-server-rendering)
* [Enable interactive Auto or WebAssembly rendering](#enable-interactive-auto-or-webassembly-rendering)

> [!NOTE]
> For the examples in this section, the example app's name and namespace is `AspNetCoreApp`.

### Add static server Razor component rendering

Add the following assets to the app, updating the `{APP NAMESPACE}` to the app's namespace as each file is added.

`Components/_Imports.razor`:

```razor
@using System.Net.Http
@using System.Net.Http.Json
@using Microsoft.AspNetCore.Components.Forms
@using Microsoft.AspNetCore.Components.Routing

> @using Microsoft.AspNetCore.Components.Web
> @using Microsoft.AspNetCore.Components.Web.Virtualization

Reviewer: ```suggestion
@using Microsoft.AspNetCore.Components.Web
@using static Microsoft.AspNetCore.Components.Web.RenderMode
@using Microsoft.AspNetCore.Components.Web.Virtualization
```

@using Microsoft.JSInterop
@using {APP NAMESPACE}
@using {APP NAMESPACE}.Components
```

`Components/Routes.razor`:

```razor
<Router AppAssembly="@typeof(Program).Assembly">
    <Found Context="routeData">

>         <RouteView RouteData="@routeData" />

Reviewer: Can we add a note about how you might want to add a default layout here that matches the rest of your app and link to the related docs on Blazor layouts?

        <FocusOnNavigate RouteData="@routeData" Selector="h1" />
    </Found>
</Router>
```

`Components/App.razor`:

```razor
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <base href="/" />
    <link rel="stylesheet" href="{APP NAMESPACE}.styles.css" />

>     <link rel="icon" type="image/png" href="favicon.png" />

Reviewer: The app might not have this specific favicon.png file.

```suggestion
```

}
```

In the ASP.NET Core project's `Program` file, add a `using` statement to the top of the file for the project's components:

```csharp
using {APP NAMESPACE}.Components;
```

Add Razor component services (`AddRazorComponents`) before the app is built (the line that calls `builder.Build()`):

```csharp
builder.Services.AddRazorComponents();
```

Add [Antiforgery Middleware](xref:blazor/security/index#antiforgery-support) to the request processing pipeline after the call to `app.UseRouting`. If there are calls to `app.UseRouting` and `app.UseEndpoints`, the call to `app.UseAntiforgery` must go between them. A call to `app.UseAntiforgery` must be placed after calls to `app.UseAuthentication` and `app.UseAuthorization`.

```csharp
app.UseAntiforgery();
```

Add `MapRazorComponents` to the app's request processing pipeline with the `App` component (`App.razor`) specified as the default root component. Place the following code before the app is run (the line that calls `app.Run`):

```csharp
app.MapRazorComponents<App>();
```

When the app is run, the `Welcome` component is accessed at the `/welcome` endpoint.

### Enable interactive server rendering

> Follow the guidance in the [Enable interactive server rendering](#enable-interactive-server-rendering) section.

Reviewer: I think you meant to link to the prior section?

```suggestion
Follow the guidance in the [Add static server Razor component rendering](#add-static-server-razor-component-rendering) section.
```

Make the following changes in the app's `Program` file.

Add a call to `AddInteractiveServerComponents` where Razor component services are added with `AddRazorComponents`:

```csharp
builder.Services.AddRazorComponents()
    .AddInteractiveServerComponents();
```

Add a call to `AddInteractiveServerRenderMode` where Razor components are mapped with `MapRazorComponents`:

```csharp
app.MapRazorComponents<App>()
    .AddInteractiveServerRenderMode();
```

Add a `Counter` component to the app with the interactive server render mode.

`Components/Pages/Counter.razor`:

```razor
@page "/counter"

> @attribute [RenderModeInteractiveServer]

Reviewer: ```suggestion
@rendermode InteractiveServer
```

<PageTitle>Counter</PageTitle>

<h1>Counter</h1>

<p role="status">Current count: @currentCount</p>

<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
    private int currentCount = 0;

    private void IncrementCount()
    {
        currentCount++;
    }
}
```

When the app is run, the `Counter` component is accessed at the `/counter` endpoint.

> ### Enable interactive Auto or WebAssembly rendering

Reviewer: I'd also like to simplify this section to make it more minimalistic:
- Clarify at the beginning that the user should first enable support for static server rendering if they don't have that already, like we do in the section on enabling interactive server rendering.
- When creating the donor project use the `--empty` option. We're just trying to enable the project structure, not include the sample pages and content.
- Remove step 2 to copy over Bootstrap and the styles. The existing app might not be using Bootstrap.
- Remove step 4. The generated project without sample pages won't have a Components folder. We'll need to add a step to add a component to the client project to demonstrate that it's working.
- Simplify step 10 given that the user has done a bunch of the described changes already as part of enabling static server rendering. Key things are to enable interactive WebAssembly support and to add the client project as an additional assembly
- Step 11 is huge! 😮 I'd like to shrink it way down to just adding another *Pages/Counter2.razor* component to the client project with `@rendermode InteractiveWebAssembly` or `InteractiveAuto`.
-

Author: Ok ... I'll put it on the new PR tho. That way, all lines go in clean for feedback.
